tools/ob_admin/backup_tool/ob_admin_backup_validation_task.cpp

```diff
@@ -0,0 +1,4275 @@
+    STORAGE_LOG(WARN, "failed to read backup info file", K(ret), K(full_path));
+  } else {
+    FOREACH_X(desc_iter, minor_tablet_to_ls_desc.tablet_to_ls_, OB_SUCC(ret))
+    {
+      const share::ObLSID &ls_id = desc_iter->ls_id_;
+      ObAdminLSAttr *ls_attr = nullptr;
+      if (OB_FAIL(ctx_->get_ls_attr(backup_set_id, ls_id, ls_attr))) {
+        if (OB_ENTRY_NOT_EXIST == ret) {
+          // encountered with post constructed logstream
+          STORAGE_LOG(WARN, "post constructed logstream", K(ret), K(ls_id));
+          ret = OB_SUCCESS;
+          if (OB_FAIL(ctx_->add_ls(backup_set_id, ls_id))) {
+            STORAGE_LOG(WARN, "failed to add ls", K(ret), K(ls_id));
+          } else if (OB_FAIL(ctx_->get_ls_attr(backup_set_id, ls_id, ls_attr))) {
+            STORAGE_LOG(WARN, "failed to get ls attr", K(ret), K(ls_id));
+          } else if (FALSE_IT(ls_attr->ls_type_ = ObAdminLSAttr::POST_CONSTRUCTED)) {
+          } else {
+            STORAGE_LOG(WARN, "post constructed logstream fixed", K(ret), K(ls_id));
+          }
+        } else {
+          STORAGE_LOG(WARN, "failed to get ls attr", K(ret), K(ls_id));
+        }
+      }
+
+      if OB_SUCC (ret) {
+        bool ls_complecity = false;
+        FOREACH_X(id_array_iter, desc_iter->tablet_id_list_, OB_SUCC(ret))
+        {
+          const common::ObTabletID &tablet_id = *id_array_iter;
+          if (tablet_id.is_ls_inner_tablet()) {
+            // record inner but not backup for post constructed
+            // just skip
+            ls_complecity = true;
+          } else if (OB_FAIL(ctx_->add_tablet(backup_set_id, ls_id, backup_minor_data_type_,
+                                              tablet_id))) {
+            STORAGE_LOG(WARN, "failed to add tablet", K(ret), K(ls_id), K(tablet_id));
+          } else if (OB_FAIL(ctx_->add_tablet(backup_set_id, ls_id, backup_major_data_type_,
+                                              tablet_id))) {
+            STORAGE_LOG(WARN, "failed to add tablet", K(ret), K(ls_id), K(tablet_id));
+          }
+        }
+        if (OB_SUCC(ret)) {
+          if (!ls_complecity) {
+            ret = OB_ERR_UNEXPECTED;
+            STORAGE_LOG(WARN, "ls complecity is false", K(ret), K(ls_id));
+          } else {
+            STORAGE_LOG(INFO, "succeed to collect tenant consistent scn tablet id", K(ret),
+                        K(backup_set_id), K(ls_id));
+          }
+        }
+      }
+    }
+    if (OB_SUCC(ret)) {
+      std::sort(backup_set_attr->minor_tablet_id_.begin(), backup_set_attr->minor_tablet_id_.end());
+      std::sort(backup_set_attr->major_tablet_id_.begin(), backup_set_attr->major_tablet_id_.end());
+      STORAGE_LOG(INFO, "succeed to collect consistent tablet id", K(ret), K(backup_set_id));
+    }
+  }
+  if (OB_FAIL(ret)) {
+    ctx_->go_abort(full_path.get_ptr(), "consistent scn data info seems corrputed");
+  }
+  return ret;
+}
+int ObAdminBackupSetMetaValidationTask::collect_minor_tablet_meta_index_()
```

Comment:
too many logic in a single function here.

Reply:
split out an inner_assign_meta_index_to_tablet_attr_